posts/2024-11-25-cloudant-with-open-liberty.adoc

```diff
@@ -0,0 +1,170 @@
+[source, java]
+----
+@ApplicationScoped
+public class CloudantProducer {
+
+    @Produces
+    public Cloudant createCloudant() {
+        return new Cloudant(“cloudant”, new BasicAuthenticator.Builder().build());
+    }
+}
+----
+
+Here is an example of using the CDI producer to inject a Cloudant client in a JAX-RS application. 
+[source, java]
+----
+@Inject
+Cloudant client;
+@POST
+@Path("/add")
+@Consumes(MediaType.APPLICATION_JSON)
+public void add(CrewMember crewMember) {
+        Document newCrewMember = new Document();			 
+        newCrewMember.put("Name",crewMember.getName());
+        newCrewMember.put("Rank",crewMember.getRank());
+        newCrewMember.put("CrewID",crewMember.getCrewID());
+
+        PostDocumentOptions createDocumentOptions =
+                    new PostDocumentOptions.Builder()
+                        .db(dbname)
+                        .document(newCrewMember)
+                        .build();
+           DocumentResult createDocumentResponse = client
+                    	.postDocument(createDocumentOptions)
+                        .execute()
+                   	    .getResult();
+}
+----
+In the above code:
+
+* `@Inject` is used to inject the Cloudant client instance provided by the CDI producer.
+* The `postDocument` method demonstrates how to create and post a document in the cloudant database using the injected CloudantClient.
+
+== Enhancing the CDI producer
+One of the advantages of using a CDI producer is that it can be tailored to your needs. For improved security,the `createCloudant` method can be enhanced to include authentication with a user name and password:
+
+[source,java]
+----
+@Produces
+public Cloudant createCloudant() {
+     String password = PasswordUtil.passwordDecode(encodedPassword);
+     BasicAuthenticator authenticator = new BasicAuthenticator.Builder()
+                .username(username)
+                .password(password)
+                .build();
+
+    return new Cloudant("cloudant", authenticator);
+}
+----
+This requires `passwordUtilities-1.0`, available as a Maven dependency:
+[source,xml]
+----
+<dependency>
+    <groupId>io.openliberty.features</groupId>
+    <artifactId>passwordUtilities-1.0</artifactId>
+    <version>18.0.0.4</version>
+</dependency>
+----
+
+You also need to enable the feature in `server.xml`:
+[source, xml]
+----
+<feature>passwordUtilities-1.0</feature>
+----
+
+By default, Cloudant communication occurs over HTTPS (SSL). However, if you're working with a custom Cloudant instance or self-signed certificates, you may need to configure your Open Liberty application to trust the custom certificate.
```

Comment:
I'm not familiar with using Cloudant with SSL, is this something that will always be necessary? If it's just for cases where self-signed certificates are used it may not be worth mentioning here, since I think that would apply to any use of self-signed certificates?